hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -177,6 +182,25 @@ private TokenAssociation validateAndBuildAutoAssociation(
             validateTrue(validAssociations, NO_REMAINING_AUTOMATIC_ASSOCIATIONS);
             validateFalse(token.hasKycKey(), ACCOUNT_KYC_NOT_GRANTED_FOR_TOKEN);
             validateFalse(token.accountsFrozenByDefault(), ACCOUNT_FROZEN_FOR_TOKEN);
+
+            final var unlimitedAutoAssociations =
+                    config.getConfigData(EntitiesConfig.class).unlimitedAutoAssociationsEnabled();
+            if (unlimitedAutoAssociations) {
+                final var topLevelPayer = handleContext.payer();
+                final var syntheticCreation = TransactionBody.newBuilder()
+                        .tokenAssociate(
+                                new Builder().account(account.accountId()).tokens(token.tokenId()));
+                final var fees = handleContext.dispatchComputeFees(
+                        syntheticCreation.build(), topLevelPayer, ComputeDispatchFeesAsTopLevel.NO);
+
+                final var storeFactory = handleContext.storeFactory();
+                final var tokenApi = storeFactory.serviceApi(TokenServiceApi.class);
+                tokenApi.chargeNetworkFee(
```

Comment:
I don't like adding `chargeNetworkFee` in multiple places inside the handler logic. It should be a part of app. But will defer to @netopyr @tinker-michaelj , what they think about this.

Reply:
I agree that this is not the right place. All fees should be calculated in the implementations of `TransactionHandler.calculateFees()`. This value is used during checks and also charged automatically in the workflow.

I am actually surprised that the `TokenServiceApi` provides this method. We should remove it IMO. Currently, it is only used in `FeeAccumulator` and we should be able to access this functionality without using `TokenServiceApi`. Same is true for the other methods that charge fees. But this is outside the scope of this PR.